core/src/main/scala/org/apache/spark/util/SparkUncaughtExceptionHandler.scala

```diff
@@ -48,11 +48,17 @@ private[spark] class SparkUncaughtExceptionHandler(val exitOnUncaughtException:
             System.exit(SparkExitCode.OOM)
           case _ if exitOnUncaughtException =>
             System.exit(SparkExitCode.UNCAUGHT_EXCEPTION)
+          case _ =>
+            // SPARK-30310: Don't System.exit() when exitOnUncaughtException is false
         }
       }
     } catch {
-      case oom: OutOfMemoryError => Runtime.getRuntime.halt(SparkExitCode.OOM)
-      case t: Throwable => Runtime.getRuntime.halt(SparkExitCode.UNCAUGHT_EXCEPTION_TWICE)
+      case oom: OutOfMemoryError =>
+        logError(s"Uncaught OutOfMemoryError in thread $thread, process halted.", oom)
```

Comment:
According to the content of try statement as of now, I guess the case we catch some exception here might probably be the case which Spark even fails to log. We may be better to do nothing but just halt.

Reply:
Thanks for the comment.  

Well the reason why for the logError is because it wasn't obvious to users or devs why the worker would just disappeared as DEAD on the UI, and there was nothing in the worker log file to tell what happened.  We couldn't find out why until we set SPARK_NO_DAEMONIZE=1 and examined the exit code.

Is there any alternative to indicate the process halted unexpectedly?

Reply:
Please take a look at the content of try statement - if you don't have any log message, then logError in try statement would have been failed already. I don't think we can succeed logging here and below in Throwable.

Reply:
So in that case, well, I guess exit code would be only the hint.

Reply:
Why would logging not succeed here?
This still halts. I think it avoids masking the OOM with a MatchError.

Reply:
Thanks for the comment.

But the `ShutdownHookManager.inShutdown()` call in the try block could also trigger another exception though, at least in theory?

But yes I agree, if the logError triggers it in the try then doing another logError in the catch won't do any good.

Reply:
I agree if it were the logError in the try block that triggers another exception then doing another logError won't do any good.  But for other cases where logError still works, it would provide better debugging messages.

Reply:
Oh OK. I agree ShutdownHookManager.inShutdown() could be...

My point is that logging here could defeat the effort of catch statement, as if logError throws another exception, the return code becomes out of control. If we really want to log it here, we may want to wrap with try-catch statement again.

Reply:
Thanks for the comment!  How about something like this?  Consider if we get there we don't really care why logError fails.

```
        try { logError(s"Uncaught OutOfMemoryError in thread $thread, process halted.", oom) }
        catch { case _: Throwable => }
        Runtime.getRuntime.halt(SparkExitCode.OOM)
```

Reply:
OK, I get it, though it is a corner case on a corner case - fatal error, then logging throws a fatal error. It might be overkill to further wrap it, but, I am not against it for complete defensiveness.

Reply:
I'd support wrapping it, as here we control the value of return code from 50 to 52 and doesn't sound ideal to have some case breaking it.

Reply:
Thanks @srowen and @HeartSaVioR, I'll make that into the next commit.